src/herder/TxSetFrame.cpp

```diff
@@ -347,207 +268,416 @@ TxSetFrame::makeFromTransactions(TxPhases const& txPhases, Application& app,
     }
 
     auto const& lclHeader = app.getLedgerManager().getLastClosedLedgerHeader();
-    // We can't use `std::make_shared` here as the constructors are protected.
-    // This may cause leaks in case of exceptions, so keep the constructors
-    // simple and exception-safe.
-    std::shared_ptr<TxSetFrame> txSet(
-        new TxSetFrame(lclHeader, validatedPhases));
-    txSet->applySurgePricing(app);
-
+    ResolvedTxSetFrame resolvedTxSet(lclHeader, validatedPhases);
+    resolvedTxSet.applySurgePricing(app);
```

Comment:
Note that this pattern occurs with other functions in resolved tx set frame class, like compute fees (basically, anything that relies on LCL should only be called on fully resolved tx sets). I'm wondering if we could strengthen the contract for ResolvedTxSetFrame, such that we can guarantee resolved tx set frames are resolved at construction. This way if we encounter ResolvedTxSetFrame in the wild, we can be sure it's resolved (so calling other functions on it won't cause invalid state)

Reply:
Also, just an idea, maybe you already considered that, but would adding a new "state" variable to the original TxSetFrame (instead of using ResolvedTxSetFrame) and allowing/disallowing certain function calls based on that state be easier to implement/reason about? (e.g. once we go unresolved->resolved, now we can call state-dependent functions on the tx set object).

Reply:
> This seems dangerous: surge pricing should only be called on fully resolved tx sets, right? So we're not actually enforcing correct ledger state here

Sure, and this tx set is fully resolved, by definition. We always have correct ledger state here because we're using the LCL to construct the tx set.

> I'm wondering if we could strengthen the contract for ResolvedTxSetFrame, such that we can guarantee resolved tx set frames are resolved at construction.

We already do that, don't we? The invariant is just enforced outside of the class because it's meaningless inside the constructor (because the constructor expects LCL itself).

> This way if we encounter ResolvedTxSetFrame in the wild, we can be sure it's resolved

This is my intention as well, and it's exactly what happens in this PR. You can't construct a `ResolvedTxSetFrame` outside of `TxSetFrame` and factory functions + `resolve` guarantee that we'll always return a correctly resolved tx set frame.

> Also, just an idea, maybe you already considered that, but would adding a new "state" variable to the original TxSetFrame (instead of using ResolvedTxSetFrame)

This was my initial thought about this, but it's much more error prone. Thanks to introducing a new class we can now identify the resolved tx set users at compile time and there is simply no way of accidentally passing unresolved tx set frame to a function that expects a resolved one.

Reply:
> Sure, and this tx set is fully resolved, by definition. We always have correct ledger state here because we're using the LCL to construct the tx set.

Well, we use `LedgerHeaderHistoryEntry` and there's no enforcement that it's the real LCL, so technically any ledger can be passed creating a "valid" resolved tx set. So could we enforce that?

Reply:
Ok, I see where you're coming from - I guess the code abused some `ResolvedTxSetFrame` functionality. I've factored out the XDR conversion, so that we only create resolved tx sets with current ledger state and they're fully processed at the end of construction (we still need to jump through some hoops to do proper initialization in `makeFromTransactions` vs `resolve`.

Reply:
Done.